### PR TITLE
Add strided layout support for to_dense

### LINKED ISSUE #59958: Make to_dense() a no-op for non-sparse Tensors
As of today, calling to_dense() on a strided Tensor raises an error
Similarly, calling to_sparse() on a sparse Tensor raises an error.

It would be nice to make these no-ops instead of errors to simplify user code.

cc @nikitaved @pearu @cpuhrsch @IvanYashchuk

### PR DESCRIPTION
**Pull request by cpuhrsch**
Fixes #59958

### CODE REVIEW

**Comment by cpuhrsch** [comment]
[deploy-linux-xenial-cuda11.3-py3.7-gcc7 / build](https://github.com/pytorch/pytorch/runs/5651503057?check_suite_focus=true#logs) fails with
```
/var/lib/jenkins/workspace/build/lib/libtorch_cuda_cu.so: undefined reference to `at::native::sparse_to_dense(at::Tensor const&, c10::optional<c10::ScalarType>)'
collect2: error: ld returned 1 exit status
caffe2/CMakeFiles/cuda_reportMemoryUsage_test.dir/build.make:111: recipe for target 'bin/cuda_reportMemoryUsage_test' failed
make[2]: *** [bin/cuda_reportMemoryUsage_test] Error 1
CMakeFiles/Makefile2:6046: recipe for target 'caffe2/CMakeFiles/cuda_reportMemoryUsage_test.dir/all' failed
make[1]: *** [caffe2/CMakeFiles/cuda_reportMemoryUsage_test.dir/all] Error 2
/var/lib/jenkins/workspace/build/lib/libtorch_cuda_cu.so: undefined reference to `at::native::sparse_to_dense(at::Tensor const&, c10::optional<c10::ScalarType>)'
collect2: error: ld returned 1 exit status
caffe2/CMakeFiles/cuda_vectorized_test.dir/build.make:147: recipe for target 'bin/cuda_vectorized_test' failed
make[2]: *** [bin/cuda_vectorized_test] Error 1
CMakeFiles/Makefile2:6000: recipe for target 'caffe2/CMakeFiles/cuda_vectorized_test.dir/all' failed
make[1]: *** [caffe2/CMakeFiles/cuda_vectorized_test.dir/all] Error 2
Makefile:140: recipe for target 'all' failed
make: *** [all] Error 2
```

**Comment by cpuhrsch** [comment]
Moving the definitions into TensorConversions.cpp solved the problem. Not sure why.

File: aten/src/ATen/native/TensorConversions.cpp [hunk 1]
```diff
@@ -240,18 +240,56 @@ Tensor to_dense_backward(const Tensor& grad, const Tensor& input_) {
...
+      !dtype.has_value(), "dtype argument is not supported by sparse_to_dense");
+  Tensor dst = at::zeros(self.sizes(), self.options().layout(kStrided));
+  return dst.add_(self);
+}
+
+Tensor sparse_csr_to_dense(
+    const Tensor& self,
+    c10::optional<ScalarType> dtype) {
+  TORCH_CHECK(
+      !dtype.has_value(), "dtype argument is not supported by sparse_to_dense");
```
Comment (pearu) [inline aten/src/ATen/native/TensorConversions.cpp:288]:
```suggestion
      !dtype.has_value(), "dtype argument is not supported by sparse_csr_to_dense");
```

File: aten/src/ATen/native/TensorConversions.cpp [hunk 2]
```diff
@@ -240,18 +240,56 @@ Tensor to_dense_backward(const Tensor& grad, const Tensor& input_) {
...
+  }
+  return tensor;
+}
+
+Tensor sparse_to_dense(
+    const Tensor& self,
+    c10::optional<ScalarType> dtype) {
+  TORCH_CHECK(
+      !dtype.has_value(), "dtype argument is not supported by sparse_to_dense");
+  Tensor dst = at::zeros(self.sizes(), self.options().layout(kStrided));
```
Comment (pearu) [inline aten/src/ATen/native/TensorConversions.cpp:280]: To support dtype for sparse cases, I'd suggest:
```suggestion
  ScalarType dtype_ = (dtype.has_value() ? dtype.value() : self.scalar_type());
  Tensor dst = at::zeros(self.sizes(), self.options().layout(kStrided).dtype(dtype_));
```

File: aten/src/ATen/native/TensorConversions.cpp (hunk 1 above)
Comment (pearu) [reply aten/src/ATen/native/TensorConversions.cpp:288]: To support dtype, see comment in `sparse_to_dense`.

File: torch/overrides.py [hunk 3]
```diff
@@ -1194,6 +1194,7 @@ def get_testing_overrides() -> Dict[Callable, Callable]:
         Tensor.tile: lambda self, *reps: -1,
         Tensor.to: lambda self, dtype, non_blocking=False, copy=False, memory_format=torch.preserve_format: -1,
         Tensor.to_dense: lambda self: -1,
+        Tensor._to_dense: lambda self: -1,
```
Comment (pearu) [inline torch/overrides.py:1197]:
```suggestion
        Tensor.to_dense: lambda self, dtype=None: -1,
        Tensor._to_dense: lambda self, dtype=None: -1,
```
?

**Review by pearu — CHANGES_REQUESTED** [review]
I have a couple of nits and a suggestion to support `sparse.to_dense(dtype)` as it appears to be straightforward.

File: torch/testing/_internal/common_utils.py [hunk 4]
```diff
@@ -2048,6 +2048,9 @@ def genSparseTensor(self, size, sparse_dim, nnz, is_uncoalesced, device, dtype):
         return x, x._indices().clone(), x._values().clone()
 
     def safeToDense(self, t):
+        # coalesce is not implemented for strided
+        if t.layout == torch.strided:
+            return t.to_dense()
         return t.coalesce().to_dense()
```
[the remark this answers is not in the document]
Comment (pearu) [reply torch/testing/_internal/common_utils.py:2054]: Interestingly, `coo.to_dense()` uses `strided.add_(coo)` which means that there is no need to coalesce the coo tensor for `to_dense`. So, `safeToDense` could be eliminated altogether.

File: aten/src/ATen/native/TensorConversions.cpp (hunk 2 above)
Comment (cpuhrsch) [reply aten/src/ATen/native/TensorConversions.cpp:280]: I like this, but it's an unrelated change and not part of the feature the issue requests. We can open a separate PR for it.

File: torch/overrides.py (hunk 3 above)
Comment (cpuhrsch) [reply torch/overrides.py:1197]: Good catch. I'm surprised the test still runs.

File: aten/src/ATen/native/TensorConversions.cpp (hunk 2 above)
Comment (pearu) [reply aten/src/ATen/native/TensorConversions.cpp:280]: Resolving via https://github.com/pytorch/pytorch/issues/74855

File: aten/src/ATen/native/TensorConversions.cpp (hunk 1 above)
Comment (pearu) [reply aten/src/ATen/native/TensorConversions.cpp]: See https://github.com/pytorch/pytorch/issues/74855

**Review by pearu — APPROVED** [review]
LGTM!

**Review by suo — APPROVED** [review]

**Pull request closed without merging**